### PR TITLE
Fix system UI font handling on macOS 11

### PR DESCRIPTION
**Pull request by vslavik**
This is an alternative to b494443 from PR #1995 that I think works (tested in Poedit). 

See what it does in the commit description, let me explain here how I arrived to it from #1995:

The b494443 (fonts fixes) commit still has problems: 

1. The font shouldn't be "SFProDisplay" (which is for large texts like titles and produces bad glyphs at small sizes like typically encountered in GUI code) nor "SFProText" (for small text), but "SFPro" — on macOS 11 that's the font dynamically choosing between the former two and the actual system font. That would be trivial to fix.

2. More troublingly, it's not enough: something as simple as `statictext->SetFont(statictext->GetFont().Bold());` doesn't work  in that the font remains with its normal weight. This is because the PostScript name changed in a way that breaks wx's expectations: in macOS 10.14, `m_postScriptName` was ".SFNSText" for the system font. In macOS 11, it is ".SFNS-Regular" — i.e. a specific weight. b494443 translates this into (with 1. applied) "SFPro-Regular", and that is then kept and used in `CreateCTFontDescriptor` regardless of the provided weight value. Making everything was stuck with the regular weight.

This is problematic to fix, because the PostScript name is part of public `wxNativeFontInfo` API, and there's really no guarantee how it's structured. For the same reason, clearing it on any weight, style etc. change is ill-advised.

I was about to propose to fix the immediate concern regarding the system font in #1995 like this instead:

```cpp
void wxNativeFontInfo::OSXAdjustPostScriptName()
{
    if ( WX_IS_MACOS_AVAILABLE(11, 0) )
    {
        if ( m_familyName == ".AppleSystemUIFont" )
            m_postScriptName.clear();
    }
}
```
(The check is for 11.0 because it only affects builds with the new SDK, which identifies the OS as 11.0, not the backward-compatible 10.16 number;  ".AppleSystemUIFont" seems new-ish too, it was ".SF NS Text" in 10.14.)

I finished with a remark that aspirationally, I think it would be best to not store `m_postScriptName` at all — use it just to create the descriptor or obtain from it when requested, but don't use it as *ground truth* as it is not being used.

But then I realized that this would break serialization in `ToString` and also some caching (which I don't quite understand). So I took a step back and did the "aspirational" thing. PostScript name is still being kept if provided by the user.

### CODE REVIEW

**Comment by csomor** [comment]
@vslavik thank's a lot for looking into this, I'll do some testing in my apps. I absolutely agree that my first attempt was in no way a definitive solution, only a first take to at least avoid getting serif fonts in the UI. Also with other script systems the system font may be completely different than the SF family.

The postscript name became more important because the font panel allowed selecting fonts variants which we don't even describe in properties. So this was the only way I found to arrive at roundtrip fidelity when serializing.

**Comment by vslavik** [comment]
> The postscript name became more important because the font panel allowed selecting fonts variants which we don't even describe in properties. So this was the only way I found to arrive at roundtrip fidelity when serializing.

I didn't know about the connection to the font panel, but did understand that this provides fidelity. I now realize my initial description was not phrased well — what I meant is not that we should remove `m_postScriptName` altogether, but that it should be one of the many *inputs* from which the `CTFontDescriptor` is created.

In other words, we shouldn't be filling it in automatically under the hood when it's not explicitly provided, because doing that subtly changes the font's specification to be a tiny bit more strict. That should be more robust handling even in absence of SF Pro issues, and that's what this patch does. That it helps with Big Sur is almost a side-effect.

**Comment by csomor** [comment]
Yes, I agree absolutely, 
I think we have a few issues to think about, things that come into my mind 

- my idea was to store the 'true' base upon which somethings was created, ie also the system ui fonts numbers if this was the way it was constructed
- I don't know however wether getting a bold variant of something constructed that way wouldn't in the end lead to the same problems, so one way would be to increase the number of font constants we have, as also windows has its constants for many of these, to offer a selection which wouldn't need modification anymore in most cases
- then the roundtrip, serialisation issue comes up, it would be great if serialization would still maintain this knowledge ... perhaps having a ps-fontname always and eventually such a constant, which would be the preferred way to reconstruct the instance ...

File: src/osx/carbon/font.cpp
Comment (vadz) [inline src/osx/carbon/font.cpp:330]: I could well be missing something, but I don't see how can this behave correctly if the PostScript name is now (almost always) empty here: these strings are cache names, so they should be unique, but now they seem to be the same for all fonts with the same style and weight. I think we should be using family/face name if PostScropt name is empty.

**Review by vadz — COMMENTED** [review]
Thanks a lot for looking at this, both your explanation and conclusion make a lot of sense to me and I'd really like to do it like this, if possible, rather than b494443989397846f3b30df54b3a1d08e64c91d5 which (no offence to Stefan intended...) looks like a hack.

But I'm worried about the cache issue, wouldn't we end up reusing the same font for everything like this?

**Comment by vslavik** [comment]
> Thanks a lot for looking at this, both your explanation and conclusion make a lot of sense to me and I'd really like to do it like this, if possible, rather than [b494443](https://github.com/wxWidgets/wxWidgets/commit/b494443989397846f3b30df54b3a1d08e64c91d5) which (no offence to Stefan intended...) looks like a hack.

I realized, and I think Stefan was referring to that above, that the hack is *still* necessary for serialization into preferences etc., so I added it back in b03fa9c3.

I agree with Stefan that longer-term ideal solution is to have more stock fonts and to serialize stock fonts as such, which would finally eliminate the need for this remapping.

> But I'm worried about the cache issue, wouldn't we end up reusing the same font for everything like this?

`GetPostScriptName()` always returns a valid PS name. The difference is that before, it would be stored the first time the font was realized and become immutable, while now it is re-obtained from the native font descriptor — which means it is updated if other characteristics (e.g. weight) change. 

My first version kept `GetPostScriptName()` as a simple getter (normally returning "") and I added `GetRealizedPostScriptName()` to make this more clear, but then I realized that some code may rely on `GetPostScriptName()` always returning a valid name even if not previously set.

**Comment by vadz** [comment]
> I agree with Stefan that longer-term ideal solution is to have more stock fonts and to serialize stock fonts as such, which would finally eliminate the need for this remapping.

Sorry, I strongly disagree with this part, it's perfectly reasonable to expect `wxFont::MakeBold()` to work for any font, stock or not, and even if we provide a stock bold font for convenience, this should still work.

> `GetPostScriptName()` always returns a valid PS name.

Ah, I've totally missed this, sorry.

> My first version kept `GetPostScriptName()` as a simple getter (normally returning "") and I added `GetRealizedPostScriptName()` to make this more clear, but then I realized that some code may rely on `GetPostScriptName()` always returning a valid name even if not previously set.

Yes, definitely. If we really wanted to have 2 functions, we'd have to add `GetExplicitlySetPSName()`, but I don't think it would be very useful.

**Comment by vslavik** [comment]
> Sorry, I strongly disagree with this part, it's perfectly reasonable to expect wxFont::MakeBold() to work for any font, stock or not,

Of course — and e86154f alone ensures that.

What is not necessarily reasonable is to *serialize*, for storage, e.g. `[NSFont boldSystemFontOfSize:13]` , as some PostScript name, which has a nasty habit of changing between macOS releases, when the OS itself doesn't use PS names natively. I.e. that's about a better alternative to the b03fa9c remapping hack.

**Comment by vadz** [comment]
@csomor Could we merge this now? AFAICS it's just waiting for your approval.

We have many unmerged wxOSX changes and I'd really like to merge them a.s.a.p. to avoid having problems later...

**Comment by csomor** [comment]
I'll do some tests today or tomorrow, which most probably will not show further problems, but just to be sure ...

**Review by csomor — APPROVED** [review]
@vslavik thanks for your work

should we could discuss the options for serializing in a follow-up conversation here, or on wx-dev ? one idea eg would be to use [familyname] if ps-name is not present and serializing back the same way, this would be backward compatible as well ...

**Comment by vadz** [comment]
Thanks for the review, I'm merging this, as I think it's sufficiently self-contained/complete for this, let's discuss the remaining issues on wx-dev.

Thanks again for the fixes!

**Comment by csomor** [comment]
Just a session link about system fonts from WWDC 2015

https://developer.apple.com/videos/play/wwdc2015/804/

**Pull request merged**
